tests/gtest/test_inline_instance.cpp

```diff
@@ -0,0 +1,318 @@
+#include "gtest/gtest.h"
+#include "coreir.h"
+#include "coreir/common/logging_lite.hpp"
+#include "coreir/ir/coreir_symbol_table.hpp"
+#include "coreir/ir/symbol_table_interface.hpp"
+#include "coreir/passes/analysis/coreirjson.h"
+
+namespace CoreIR {
+namespace {
+
+class InlineInstance : public ::testing::Test {
+ public:
+  InlineInstance() = default;
+
+  void SetUp() override {
+    ctx = newContext();
+    ctx->setDebug(true);
+    if (!loadFromFile(ctx, "srcs/simple_hierarchy.json", &top)) ctx->die();
+    assert(top != nullptr);
+    ctx->setTop(top->getRefName());
+  }
+
+  void TearDown() override {
+    if (table != nullptr) std::cout << table->json() << std::endl;
+    deleteContext(ctx);
+  }
+
+ protected:
+  Context* ctx = nullptr;
+  Module* top = nullptr;
+  SymbolTableInterface* table = nullptr;
+  SymbolTableSentinel* const empty = symbolTableEmptySentinel();
+  SymbolTableSentinel* const inlined = symbolTableInlinedInstanceSentinel();
+};
+
+template <typename VariantType, typename T> void VariantExpectEq(
+    const VariantType& container, const T& expected) {
+  try {
+    auto typed = std::get<T>(container);
+    EXPECT_EQ(typed, expected);
+  } catch (const std::bad_variant_access&) {
+    EXPECT_TRUE(false) << "VariantExpectEq failed with wrong type";
+  }
+}
+
+TEST_F(InlineInstance, Basic) {
+  // Grab handles to M1.i2, M1.i2.i3.
+  const auto inst_M1_i2 = top->getDef()->getInstances().at("i2");
+
+  // Inline M1.i2.
+  inlineInstance(inst_M1_i2);
+
+  // Check that M1 has only an instance of M3 named i2$i3.
+  EXPECT_EQ(top->getDef()->getInstances().size(), 1);
+  for (auto [name, inst] : top->getDef()->getInstances()) {
+    if (name == "i2$i3") {
+      EXPECT_EQ(inst->getModuleRef()->getName(), "M3");
+    } else {
+      ASSERT_TRUE(false) << "Unexpected instance";
+    }
+  }
+}
+
+TEST_F(InlineInstance, SymbolTableBasic) {
+  // Grab handles to M1.i2.
+  const auto inst_M1_i2 = top->getDef()->getInstances().at("i2");
+  // Inline M1.i2.
+  inlineInstance(inst_M1_i2);
+
+  std::string flattentypes_str = "flattentypes";
+  ctx->runPasses({"flattentypes", "verilog"}, {});
+
+  table = ctx->getPassManager()->getSymbolTable();
+  table->getLogger()->finalize();
+  // Check module names (sanity check).
+  EXPECT_EQ(table->getModuleName("M1"), "M1");
+  // This won't work since there are no longer instances of M2:
+  //   EXPECT_EQ(table->getModuleName("M2"), "M2");
```

Comment:
Why does the symbol table remove M2 if an instance of M2 is inlined? what if there are other instances of M2 in other modules?

Reply:
then it won't be removed. its just because the module names are only inserted into the symbol table during verilog generation which is an instance graph pass. if a module is "unused" in that, then it won't appear in the symbol table.